[Video] Import/export improvements and fixes - including multi-episode file support (including blurays).

78andyp opened the pull request:
## Description
<!--- Provide a general summary of your change in the Pull Request title above -->
<!--- Describe your change in detail here. -->

Allows the import/export of multi-episode files (which includes blurays).
Allows nfos to be override nfos.

Main changes

- For multi-episode files (media or bluray) this PR appends -SxxEyy to the filename and an .nfo is created for each episode. For bluray files these files are created in the BDMV directory eg. index-S01E01.nfo

NB art is placed there as well (with the same name and art type appended - eg. `index-SxxEyy-thumb.jpg`) - this is in contrast to other art that is placed in the basepath for blurays. For multi-episode media files the nfo files and art are all in the same directory (as currently). This fixes the art issue below and also ensures the XMLs are compliant as they only contain one `<episodedetails>` tag.

- Also introduces an override tag for nfos - so for example, if  you scrape a directory using an online scraper but want to specify a playlist then create an nfo named (for example) index-S01E01.nfo containing:
```
<episodedetails override="true">
  <playlist>140</playlist>
</episodedetails>
```

## Motivation and context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

Whilst adding bluray support I found a few issues with current implementation. The movie related ones have now moved to the bluray version PR.

**Failure to save/import art for multi-episode media files and non-compliant XML**

Create a multi-episode file

![image](https://github.com/user-attachments/assets/103da866-3ad5-448c-abd8-745ba414cb86)

![image](https://github.com/user-attachments/assets/50162436-5724-4776-a5c1-ce69f660d34c)

Set content TV Shows -> Local scraper
XML not compliant (as multiple `<episodedetails>`)

![image](https://github.com/user-attachments/assets/da5d3a6b-ec68-4695-8612-b32a6e2edde4)

The issue is art is only saved for one episode.

Set content TV Shows-> Local Scraper

![image](https://github.com/user-attachments/assets/2f872f25-d9c1-41e5-9477-f3ce7199d150)

Note the thumb is the same.

## How has this been tested?
<!--- Please describe in detail how you tested your change. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

Creating a sample source containing a mixture of media files, a DVD VIDEO_TS, a bluray ISO and BDMV.
Export, delete database and add source.
Examine the movie, paths, files and versions tables.

Note the movie fixes (other than images being written in the wrong directory and export not completing) have moved into another PR.

Episodes

![image](https://github.com/user-attachments/assets/1d8f8437-c31c-4cc1-b5a5-8559091f619b)

![image](https://github.com/user-attachments/assets/a6cc5a8a-9db5-4b5c-b7fd-3eb8354c1c61)

![image](https://github.com/user-attachments/assets/41fe2486-fb29-4d03-8e7b-d5311c979643)

After Export -> Delete -> Add Source

![image](https://github.com/user-attachments/assets/d7e42d5b-7c69-4cc9-adcf-55c93338bc93)

![image](https://github.com/user-attachments/assets/1d002c58-0b09-4d78-afdc-30f84e474674)

![image](https://github.com/user-attachments/assets/66a22dda-6750-4741-a2c5-7950bcdf3bf9)

## What is the effect on users?
<!--- Summarize the effect of this change on Kodi end-users. -->
<!--- If the PR does not have a noticeable impact (e.g., if it only changes documentation), -->
<!--- just leave it empty. Put in more detail the bigger the impact is. -->
<!--- This section may be used for automatic creation of release notes. -->

## Screenshots (if appropriate):

## Types of change
<!--- What type of change does your code introduce? Put an `x` with no space in all the boxes that apply like this: [X] -->
- [X] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [X] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **Student submission** (PR was done for educational purposes and will be treated as such)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` with no space in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [X] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [X] All new and existing tests passed

78andyp commented:
Hi @CrystalP @garbear - this should be straightfoward as well. Thanks :)

CrystalP commented:
Not that straightforward, this PR adds a new tag <relativefilename> to nfo files, which has to be documented.
So please provide some explanations and usage examples.

78andyp commented:
> Not that straightforward, this PR adds a new tag `<relativefilename>` to nfo files, which has to be documented. So please provide some explanations and usage examples.

Apologies for not getting back to you sooner on this.
I've changed the code do that's not necessary any more.
There just needed to be a way to communicate the playlist through the export->import process. I can do that with the existing <file> tag.

CrystalP commented:
Using the <file> tag that's normally in the single file export with a different meaning in the export to separate files is worse as far as I am concerned and still is a new tag for .nfo files, to be documented because they're public files, unlike the videodb.xml of the single file export which can be considered an opaque file of Kodi.

The export to separate files doesn't work for the menu (at least Bluray folder) - no .nfo created.

unrelated to this PR - a movie stopped on the menu fails to play again ("this file is no longer available")

78andyp commented:
Aarrgghh. Menus - that should have been fixed. I've been moving stuff around a bit so probably broke something - I'll check again - thanks for pointing out.

What do  you think is best for recording the bluray:// path in the single xml. The issue is - as the bluray:// path is not derived at scan time but at play time the deafult action of export/import (when paths are disabled) means it's lost - so irrespective of the save paths setting it needs to be recorded. In some ways, the above is correct - as it is the filename - but I can see where you're coming from. Are you OK if I revert it to what I had before - a new tag and just add that to the wiki (with my new found wiki editing powers!!)??

CrystalP commented:
As far as I can tell exporting to a single file already populates the file, path, filenameandpath and basepath tags, which include the playlist or menu.
The problem is the export to separate .nfo files. A new tag might work in the short term but that design may not be enough a bit later when considering other things you want to achieve for Bluray, like multiple movies, multiple versions, multiple episodes, extras on one disc. With the current rules for nfo name and location, there can only be one nfo per disc and that won't be enough...
A solution for the bigger problem could guide the solution used in this PR, as a subset of the big solution.

78andyp commented:
Good point. Will review.

78andyp commented:
@KarellenX - are you able to test this?? There is a windows test build at https://mirrors.kodi.tv/test-builds/windows/win64/KodiSetup-20250523-d26a6df8-blurayfixes3-x64.exe. It doesn't matter if you can't test the bluray bits, they ended up being a relatively small part of this - I'm more concerned if there are any regressions and I know you use nfo files extensively. And if you do have any multi-episode files or verisons that would be a bonus.

KarellenX commented:
@78andyp 

I have done some testing.
After your recent fix to enable Update Library of Movies and TV Shows, that component now works correctly.

Next I tested multi-episode files. I thought there was a problem there until further testing revealed there is a change in behaviour. (Yea, yea, I should have read the first post in more detail  😆  )

I originally tested using an existing tv show with existing nfo files.
When I scanned this in, all the multi-episode entries inherited the metadata from the first episode in the batch.

![002077](https://github.com/user-attachments/assets/6f567723-a92f-4194-9b0b-106eb50bdda5)

I then further tested by scraping the tv show first. This scraped correctly and all multi-episodes had correct data.
I then exported the library and artwork.
After the export, I noticed the exports were different.

Original multi-episode looks like this...
```
S01E01E02E03E04.mkv
S01E01E02E03E04.nfo
S01E01E02E03E04-thumb.jpg
```

Modified export now looks like this...
```
S01E01E02E03E04.mkv
S01E01E02E03E04-S01E01.nfo
S01E01E02E03E04-S01E01-thumb.jpg
S01E01E02E03E04.mkv
S01E01E02E03E04-S01E02.nfo
S01E01E02E03E04-S01E02-thumb.jpg
S01E01E02E03E04.mkv
S01E01E02E03E04-S01E03.nfo
S01E01E02E03E04-S01E03-thumb.jpg
S01E01E02E03E04.mkv
S01E01E02E03E04-S01E04.nfo
S01E01E02E03E04-S01E04-thumb.jpg
```

I think this is great as it fixes a long-standing problem of not being able to assign different local episode images to the multi-episode file.
The only drawback I see is that users with existing nfo files will need to modify their local files for this new process to work correctly. If they scan the existing single file nfo file, then run an export, the new format nfo files are created, but containing the same data across all the nfo files. Any workaround for that? I guess I'll just answer those questions on the forum when they pop-up.

Next up, I'll test the bluray nfo files, but I am not really sure how to test that. I read your statement below, but how do I allocate, say, episode 1 to the correct m2ts or pls file? Maybe ping me on slack?
```For bluray files these files are created in the BDMV directory and art is placed there as well (with the same name and art type appended - eg. index-SxxEyy-thumb.jpg)```

78andyp commented:
Thanks for the testing. I'll explain blurays here for reference (and it would be useful to have it all in one place for updating the wiki as well).

You are right about existing multi-episode nfos. I made a conscious decision to remove the exisitng multiple <episodedetails> entries in a single nfo because:
- It's not XML compliant
- I figured that multi-episode files are probably rare
- Probably a smaller subset of people then use nfos over time to rebuild a library (ie. export them and a period of time later re-import them)
- We are still pre-alpha so there is plenty of time for the change to be baked in and for tools that generate nfos to be updated.
However they are all assumptions and happy to be challenged on them.

We may need to handle the situation better though - I'll ping you on slack with some thoughts. I wonder if just ignoring the multi episode nfo would be best (and most obvious to the end user).

Right... blurays. Two episode examples - one with an ISO and one with folders.

First looking at ISO - folder structure:
![image](https://github.com/user-attachments/assets/73bde5f5-2440-4a14-97e7-eb636a0d0e65)

Files:
![image](https://github.com/user-attachments/assets/97d018ed-9bfc-4983-85c8-a9a4428a310a)

Example nfo (S1E02)
https://paste.kodi.tv/epiyujipox

The addition is the `<playlist>102</playlist>` tag. This only appears (currently) if the episode has actually been played at least once as the playlist is selected via the simple menu at play time.

Now BDMV folders:
![image](https://github.com/user-attachments/assets/68e04cd7-0383-4e70-b522-e10251bdeabb)

Files (in BDMV):
![image](https://github.com/user-attachments/assets/ee8a5b2b-265a-4c07-8344-8b0c509a5aa5)

Example nfo (S01E01)
https://paste.kodi.tv/lunubalife

This time no `<playlist>` as it hasn't been played yet.

78andyp commented:
Jenkins build this please

78andyp commented:
The episodes part of this has been well tested - I'm going ot merge it (I'll remove the movie import/export commits and put them in the movie versions PR prior to commit) in the next couple of days unless there are any objections.

78andyp committed:
Fix an issue with movie exports (pending future movie specific nfo update).

- Use local dataset as m_pDS can get overwritten in GetFile/GetPath which is used in called functions for bluray:// paths (meant some movies are not exported).

78andyp committed:
Art file related refactoring and tests.

78andyp committed:
Export multi-episode files (including blurays).

** This changes .nfo/art file naming for separate file exports from a multi-episode media file.

Currently if a file contains multiple episodes, a single .nfo is created with multiple root <episodedetails> xml tags. This is not XML compliant.

As a side effect - when exporting to separate files, it means only art for the first episode is saved (named after the multi-episode file).

This commit changes this so that -SxxEyy is appended to the filename and an .nfo is created for each episode.

This means that the xml is compliant and episode art is preserved.

This commit uses a mapping function to resolve bluray:// paths to base paths to establish which episodes (played - where the file name is changed to bluray:// and contains the playlist - and unplayed) belong in which base file.

78andyp committed:
Import multi-episode files (including blurays).

Ensures the bluray:// paths are correctly translated to/from real file paths.

Also remove redundant code in CNfoFile as episodes now have individual .nfo files.

78andyp committed:
Allow override nfos.

78andyp commented on xbmc/video/VideoDatabase.cpp:
```diff
@@ -3807,62 +3807,89 @@ void CVideoDatabase::GetEpisodesByFileId(int idFile, std::vector<CVideoInfoTag>&
   catch (...)
   {
-    CLog::LogF(LOGERROR, "File Id {} failed", idFile);
+    CLog::LogF(LOGERROR, "Failed to fetch episodes");
```
Done.

78andyp commented:
Unless there are any objections, I'm going to merge this is a couple of days - it's been around for a while and tested in bluray-external slack.

78andyp commented on xbmc/video/VideoInfoScanner.cpp:
```diff
@@ -140,10 +140,11 @@ void AddLocalItemArtwork(KODI::ART::Artwork& itemArt,
-                               (caseSensitive
-                                    ? StringUtils::StartsWith(candidate, baseFilename)
-                                    : StringUtils::StartsWithNoCase(candidate, baseFilename))};
+    std::string baseFilename{initialBaseFilename};
```
Done

78andyp commented on xbmc/video/tags/VideoTagLoaderNFO.cpp:
```diff
@@ -20,6 +20,7 @@
 #include "utils/log.h"
 #include "video/VideoInfoTag.h"
 
+#include <ranges>
```
Fair enough - it looks like  `std::ranges::distance`  uses `<iterator>`

78andyp commented on xbmc/video/tags/VideoTagLoaderNFO.cpp:
```diff
@@ -167,36 +185,39 @@ std::string CVideoTagLoaderNFO::FindNFO(const CFileItem& item,
+                                  if (tag->m_iSeason >= 0 && tag->m_iEpisode >= 0)
+                                  {
+                                    std::string path{nfoItem->GetPath()};
+                                    size_t period = path.find_last_of("./\\");
+                                    if (period != std::string::npos && path[period] == '.')
+                                      path.erase(period);
```
Funnily enough I tried RemoveExtension() doesn't actually remove extensions from all files - there are some it doesn't, but agree this would achieve the same effect.

78andyp commented on xbmc/utils/ArtUtils.cpp:
```diff
@@ -235,6 +241,21 @@ std::string GetLocalArtBaseFilename(const CFileItem& item, bool& useFolder)
+  {
+    // Note this is an exception to the optical media rule above - episode art files will be stored alongside the episode .nfo
+    const CVideoInfoTag* tag{item.GetVideoInfoTag()};
+    if (tag->m_iSeason > -1 && tag->m_iEpisode > -1)
```
DOne

CrystalP commented:
I don't see an answer to my comment on 26963 about backwards compatibility that really belonged in this PR:
"I didn't runtime test and may have misunderstood, but does this mean that old style shared multi-episode tbn won't be recognized anymore?
I think it's OK to give priority to the new more specific S01E01E02-S01E01.tbn format, but I think the "legacy" S01E01E02.tbn should still be recognized as a fallback."
@KarellenX maybe you have an opinion about this.

KarellenX commented:
>I don't see an answer to my comment on 26963 about backwards compatibility and fallback that really belonged in this PR:
"I didn't runtime test and may have misunderstood, but does this mean that old style shared multi-episode tbn won't be recognized anymore?
I think it's OK to give priority to the new more specific S01E01E02-S01E01.tbn format, but I think the "legacy" S01E01E02.tbn should still be recognized as a fallback."
@KarellenX maybe you have an opinion about this.

@CrystalP 

Backwards compatibility- I don't think it that important. Yep, there will be questions on the forum about why artwork is not working, and we will point them to the new method which most, if not all, users will think an improvement.

I thought that the .tbn extension was deprecated.
